### PR TITLE
kubelet: cgroupv2 disable memcg notifications

### LINKED ISSUE #106331: kubelet kernelMemcgNotification are not supported on cgroupv2
### What happened?

Start kubelet on a cgroupv2 image with `kernelMemcgNotification` kubelet config option or the flag `--kernel-memcg-notification=true` enabled. Kubelet will periodically log something like:

```
kubelet[2571]: I1109 00:16:36.071429    2571 eviction_manager.go:262] "Eviction manager: failed to update notifier" notifier="hard memory eviction threshold" err="no such file or directory"
```

This is because the kernel memcg notifications relies on kubelet's `threshold_notifier` which uses evenfd on `cgroup.event_control` file which is cgroupv1 specific and not supported on cgroupv2.

https://github.com/kubernetes/kubernetes/blob/3f0a634ef78b6df33075e54b0f89e465e88049c4/pkg/kubelet/eviction/threshold_notifier_linux.go#L55-L59

### What did you expect to happen?

Kubelet will disable kernel memcg notifications on cgroupv2 if they are unsupported and consider long term support for more cgroupv2 "compatible" ways to monitor for memory pressure.

The closest analogous cgroupv2 concept is using inotify notifications on `memory.events` file and or combination with PSI pressure. This requires `memory.low` to be set on pod cgroup which requires the MemoryQOS https://github.com/kubernetes/enhancements/issues/2570 feature to be enabled

### How can we reproduce it (as minimally and precisely as possible)?

Start kubelet on a cgroupv2 image with `kernelMemcgNotification` enabled.

### PR DESCRIPTION
**Pull request by bobbypage**
#### What type of PR is this?

/kind bug 

#### What this PR does / why we need it:

The kubelet memory notifier relies on reading
`cgroup.event_control` which is unsupported on cgroupv2. For now, let's
disable the memory notifier on cgroupv2.

#### Which issue(s) this PR fixes:

Fixes #106331

#### Special notes for your reviewer:

As a more short term solution, I propose to disable kernel memcg notifications on kubelet for cgroupv2 since it's currently not supported. Longer term, we should consider how we adapt kubelet eviction as a whole to take advtange of new cgroupv2 features. 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by bobbypage** [comment]
/hold

To get some feedback and perform some testing

**Review by giuseppe — APPROVED** [review]
LGTM

**Comment by bobbypage** [comment]
Tested as follows, started a cluster on GCE with ubuntu cgroupv2 image with and without this change:

```
# ubuntu-2110 "Impish Indri" has cgroupv2 enabled
UBUNTU_IMAGE=$(gcloud compute images list --project="ubuntu-os-cloud" --no-standard-images --format=json | jq -r '.[] | select(.family=="ubuntu-2110") | .name')

KUBE_UBUNTU_INSTALL_CONTAINERD_VERSION=v1.5.7 \
  KUBE_UBUNTU_INSTALL_RUNC_VERSION=v1.0.2 \
  KUBE_MASTER_OS_DISTRIBUTION=ubuntu \
  KUBE_GCE_MASTER_IMAGE=${UBUNTU_IMAGE} \
  KUBE_GCE_MASTER_PROJECT=ubuntu-os-cloud \
  KUBE_NODE_OS_DISTRIBUTION=ubuntu \
  KUBE_GCE_NODE_IMAGE="${UBUNTU_IMAGE}" \
  KUBE_GCE_NODE_PROJECT=ubuntu-os-cloud \
  KUBE_KUBELET_EXTRA_ARGS="--kernel-memcg-notification=true" kubetest \
    --build \
    --up \
    --provider=gce \
    --gcp-node-image=ubuntu \
    --gcp-master-image=ubuntu \
    --gcp-zone=us-central1-c \
    --gcp-project=${GCP_PROEJCT} \
    --gcp-nodes=1
```

Without this change (failed to update notifier due to reading non-existent `cgroup.event_control` file)
```
$ journalctl -u kubelet | grep "eviction"
Nov 11 00:36:37 e2e-test-porterdavid-minion-group-wsqk kubelet[6645]: I1111 00:36:37.426226    6645 memory_threshold_notifier.go:101] "Eviction manager: setting notifier to capacity" notifier="hard memory eviction threshold" capacity="9139312Ki"
Nov 11 00:36:37 e2e-test-porterdavid-minion-group-wsqk kubelet[6645]: I1111 00:36:37.426278    6645 eviction_manager.go:262] "Eviction manager: failed to update notifier" notifier="hard memory eviction threshold" err="no such file or directory"
Nov 11 00:36:37 e2e-test-porterdavid-minion-group-wsqk kubelet[6645]: I1111 00:36:37.426298    6645 memory_threshold_notifier.go:101] "Eviction manager: setting notifier to capacity" notifier="hard allocatable memory eviction threshold" capacity="7373584Ki"
Nov 11 00:36:37 e2e-test-porterdavid-minion-group-wsqk kubelet[6645]: I1111 00:36:37.426325    6645 eviction_manager.go:262] "Eviction manager: failed to update notifier" notifier="hard allocatable memory eviction threshold" err="no such file or directory"
Nov 11 00:36:37 e2e-test-porterdavid-minion-group-wsqk kubelet[6645]: I1111 00:36:37.426572    6645 eviction_manager.go:328] "Eviction manager: no resources are starved"
```

and with this change (no such errors):
```
Nov 11 01:07:20 e2e-test-porterdavid-minion-group-05km kubelet[4160]: I1111 01:07:20.364273    4160 memory_threshold_notifier.go:101] "Eviction manager: setting notifier to capacity" notifier="hard allocatable memory eviction threshold" capacity="7371440Ki"
Nov 11 01:07:20 e2e-test-porterdavid-minion-group-05km kubelet[4160]: I1111 01:07:20.365310    4160 eviction_manager.go:328] "Eviction manager: no resources are starved"
Nov 11 01:07:30 e2e-test-porterdavid-minion-group-05km kubelet[4160]: I1111 01:07:30.366155    4160 eviction_manager.go:237] "Eviction manager: synchronize housekeeping"
Nov 11 01:07:30 e2e-test-porterdavid-minion-group-05km kubelet[4160]: I1111 01:07:30.386698    4160 memory_threshold_notifier.go:101] "Eviction manager: setting notifier to capacity" notifier="hard memory eviction threshold" capacity="9134780Ki"
Nov 11 01:07:30 e2e-test-porterdavid-minion-group-05km kubelet[4160]: I1111 01:07:30.386719    4160 memory_threshold_notifier.go:101] "Eviction manager: setting notifier to capacity" notifier="hard allocatable memory eviction threshold" capacity="7371440Ki"
Nov 11 01:07:30 e2e-test-porterdavid-minion-group-05km kubelet[4160]: I1111 01:07:30.386949    4160 eviction_manager.go:328] "Eviction manager: no resources are starved"
```

File: pkg/kubelet/eviction/threshold_notifier_linux.go [hunk 1]
```diff
@@ -46,6 +47,14 @@ var _ CgroupNotifier = &linuxCgroupNotifier{}
 // NewCgroupNotifier returns a linuxCgroupNotifier, which performs cgroup control operations required
 // to receive notifications from the cgroup when the threshold is crossed in either direction.
 func NewCgroupNotifier(path, attribute string, threshold int64) (CgroupNotifier, error) {
+	// cgroupv2 does not support monitoring cgroup memory thresholds using cgroup.event_control.
+	// Instead long term, on cgroupv2 kubelet should rely on combining usage of memory.low on root pods cgroup with inotify notifications on memory.events and or PSI pressure.
+	// For now, let's return a fake "disabled" cgroup notifier on cgroupv2.
+	// https://github.com/kubernetes/kubernetes/issues/106331
+	if libcontainercgroups.IsCgroup2UnifiedMode() {
```
Comment (249043822) [inline pkg/kubelet/eviction/threshold_notifier_linux.go:54]: Should we not start the notifier at all in eviction_manager while enable kernelMemcgNotification 
 but on cgroupv2 ?https://github.com/kubernetes/kubernetes/blob/c98b388a847f84019609f6422b4ab00b89c4603c/pkg/kubelet/eviction/eviction_manager.go#L184

Comment (bobbypage) [reply pkg/kubelet/eviction/threshold_notifier_linux.go:54]: that would work too, but I choose to disable it in the `NewCgroupNotifier`, so calling `Start()` will be a no-op. I guess there's not big difference on where exactly to disable it.

**Comment by pacoxu** [comment]
@bobbypage should we mention it in the kubelet configuration descriptions

Field | Description
-- | --
kernelMemcgNotificationbool | kernelMemcgNotification, if set, instructs the kubelet to integrate with the kernel memcg notification for determining if memory eviction thresholds are exceeded rather than polling. Default: false


https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1/

**Pull request merged**